cross-compilation regression on mirage

Our mirage CI run is broken. The linking step fails with:

```
/usr/bin/ld: /home/runner/work/dune/dune/mirage/_build/solo5/duniverse/Zarith/../../../install/default/lib/gmp//libgmp.a(assert.o): in function `__gmp_assert_header':
assert.c:(.text+0x36): undefined reference to `__fprintf_chk'
/usr/bin/ld: /home/runner/work/dune/dune/mirage/_build/solo5/duniverse/Zarith/../../../install/default/lib/gmp//libgmp.a(assert.o): in function `__gmp_assert_fail':
assert.c:(.text+0x9a): undefined reference to `__fprintf_chk'
/usr/bin/ld: /home/runner/work/dune/dune/mirage/_build/solo5/duniverse/Zarith/../../../install/default/lib/gmp//libgmp.a(assert.o): in function `__gmp_assert_header':
assert.c:(.text+0x6c): undefined reference to `__fprintf_chk'
/usr/bin/ld: /home/runner/work/dune/dune/mirage/_build/solo5/duniverse/Zarith/../../../install/default/lib/gmp//libgmp.a(memory.o): in function `__gmp_default_allocate':
memory.c:(.text+0x42): undefined reference to `__fprintf_chk'
/usr/bin/ld: /home/runner/work/dune/dune/mirage/_build/solo5/duniverse/Zarith/../../../install/default/lib/gmp//libgmp.a(memory.o): in function `__gmp_default_reallocate':
memory.c:(.text+0x95): undefined reference to `__fprintf_chk'
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:
```

I bisected that to #8740. I'll try to isolate a repro case.

--- a/src/dune_rules/expander.ml
+++ b/src/dune_rules/expander.ml
@@ -332,8 +332,12 @@ let expand_lib_variable t source ~lib ~file ~lib_exec ~lib_private =
                     |> String.quoted)
              ])
     else (
-      let artifacts = if lib_exec then t.lib_artifacts_host else t.lib_artifacts in
-      file_of_lib artifacts (Context.host t.context) ~loc ~lib ~file)
+      let artifacts, context =
+        if lib_exec
+        then t.lib_artifacts_host, Context.host t.context
+        else t.lib_artifacts, t.context
+      in
+      file_of_lib artifacts context ~loc ~lib ~file)
   in
   let p =
     let open Memo.O in

--- a/test/blackbox-tests/test-cases/github9199.t
+++ b/test/blackbox-tests/test-cases/github9199.t
@@ -34,4 +34,4 @@ See #9199.
   $ cat _build/host/target
   ../install/host/lib/somelib/somelib.a
   $ cat _build/cross/target
-  ../install/host/lib/somelib/somelib.a
+  ../install/cross/lib/somelib/somelib.a

